### PR TITLE
Coinbase spends will be removed after reorg

### PR DESCRIPTION
**Pull request by codingupastorm**

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs [hunk 1]
```diff
@@ -65,6 +75,28 @@ private async Task AddBackToMempoolAsync(Block block)
             }).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// If there are any transactions in the mempool that depend on transactions no longer in the chain, remove them.
+        /// </summary>
+        private async Task RemoveInvalidTransactionsAsync(Block block)
+        {
+            // Invalid transactions would have spent the coinbase. The other transactions can be put back into the mempool and be fine.
+            uint256 coinbaseId = block.Transactions[0].GetHash();
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs]: For safety it's probably best to check that this really is a coinbase

File: src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs [hunk 2]
```diff
@@ -65,6 +75,28 @@ private async Task AddBackToMempoolAsync(Block block)
...
+
+            await this.mempoolLock.WriteAsync(async () =>
+            {
+                foreach (TxMempoolEntry mempoolEntry in this.mempool.MapTx.SpendsCoinbase.ToList())
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs]: I agree that this is important logic that needs to be added, but on further consideration I think it is incomplete.

Consider the case where we are on fork A and reorg to fork B. Now all of the transactions on fork A that are beyond the fork point need to be re-evaluated. This is particularly true in the case of multi-block reorgs: a transaction may be put back into the pool when the tip is disconnected, then its own input(s) maybe be removed when (tip-1) is disconnected, etc.

At minimum the following would need to be re-evaluated, in my view:
a. Existence of inputs (they may have been reorged away; for example the coinbase, such as in your PR)
b. Coinbase/stake maturity requirements (it is possible that fork B is shorter than A such that these requirements are no longer met)
c. Similarly to (b), locktime requirements
d. Possibly others

For now this PR can be merged standalone, but I would suggest adding a TODO as a reminder to do the rest.

Comment (codingupastorm) [reply src/Stratis.Bitcoin.Features.MemoryPool/BlocksDisconnectedSignaled.cs]: Thanks, made the changes as suggested.

File: src/Stratis.SmartContracts.IntegrationTests/ContractPoAStressTests.cs [hunk 3]
```diff
@@ -53,5 +61,75 @@ public void MaximumCreateTransactionsInABlock()
...
+            var node2 = this.mockChain.Nodes[1];
+            var node3 = this.mockChain.Nodes[2];
+
+            int startingHeight = node1.CoreNode.FullNode.ChainIndexer.Height;
+
+            // Nodes are syncing together...
+            this.mockChain.MineBlocks(1);
+
+            // Node 1 loses connection to the others
+            foreach (var peer in node1.CoreNode.FullNode.ConnectionManager.ConnectedPeers.ToList())
```
Comment (fassadlr) [inline src/Stratis.SmartContracts.IntegrationTests/ContractPoAStressTests.cs]: You can replace this `foreach` and the `WaitLoop` with one line:

```
TestHelper.DIsconnectAll(node1)
```

File: src/Stratis.SmartContracts.IntegrationTests/ContractPoAStressTests.cs [hunk 4]
```diff
@@ -53,5 +61,75 @@ public void MaximumCreateTransactionsInABlock()
...
+                }
+            });
+            Transaction tx = node1.CoreNode.FullNode.Network.CreateTransaction(response2.Hex);
+            Assert.Equal(unspents[1].Transaction.Id, tx.Inputs[0].PrevOut.Hash);
+            node2.WaitMempoolCount(1);
+            node3.WaitMempoolCount(1);
+
+            // Other node mines far ahead
+            node1.CoreNode.MineBlocksAsync(5).GetAwaiter().GetResult();
+            Assert.True(node2.CoreNode.FullNode.ChainIndexer.Height == startingHeight + 3);
```
Comment (fassadlr) [inline src/Stratis.SmartContracts.IntegrationTests/ContractPoAStressTests.cs]: You can replace these lines with `TestHelper.IsNodeSyncedAtHeight(node2, startingHeight + 3);`

(If you wanna)

File: src/Stratis.SmartContracts.IntegrationTests/ContractPoAStressTests.cs [hunk 5]
```diff
@@ -53,5 +61,75 @@ public void MaximumCreateTransactionsInABlock()
...
+            // Reconnect nodes.
+            TestHelper.Connect(node1.CoreNode, node2.CoreNode);
+            TestHelper.Connect(node1.CoreNode, node3.CoreNode);
+
+            // 2 and 3 will reorg to 1's chain.
+            TestBase.WaitLoop(() => node2.CoreNode.FullNode.ChainIndexer.Height == startingHeight + 6);
+            TestBase.WaitLoop(() => node3.CoreNode.FullNode.ChainIndexer.Height == startingHeight + 6);
+
+            // Lets give some time to the nodes to try and sort themselves out.
+            Thread.Sleep(5_000);
```
Comment (fassadlr) [inline src/Stratis.SmartContracts.IntegrationTests/ContractPoAStressTests.cs]: Generally the TestHelper's `AreNodesSynced` method should wait until all the nodes are on the same height.

**Pull request merged**